File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow=head;
        ListNode* fast=head;
        while(fast!= NULL&& fast->next!= NULL){
            slow=slow->next;
            fast=fast->next->next; // two step ahead
        }
    
    // reverse array2
    ListNode* curr= slow->next; // node dau day 2
    ListNode *prev= NULL;
    slow->next=NULL;
    while(curr !=NULL){
        ListNode* tmp= curr->next; 
        curr->next=prev; 
        prev=curr;
        curr=tmp;
    }

    // 3. merge
    ListNode* l1 = head;
    ListNode* l2 = prev;// tro den node cuoi 

    while(l2){
        ListNode* n1 = l1->next;
        ListNode* n2 = l2->next;

        l1->next = l2;
        l2->next = n1;

        l1 = n1;
        l2 = n2;
    }
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# xóa phần tử thứ n từ cuối
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        cur =head
        nodes=[]
        while cur:
            nodes.append(cur)
            cur=cur.next
        delindex=len(nodes)-n
        if delindex==0:
            return head.next
        nodes[delindex-1].next=nodes[delindex].next
        return head
        